xdem/terrain.py

```diff
@@ -230,3 +259,156 @@ def hillshade(
+        * NaNs and infs are filled with the median of the finites in the matrix, possibly affecting the fit.
+        * The X and Y resolution needs to be the same. It does not work if they differ.
+
+    :param dem: The 2D DEM to be analyzed (3D DEMs of shape (1, row, col) are not supported)
+    :param resolution: The X/Y resolution of the DEM.
+
+    :raises ValueError: If the inputs are poorly formatted.
+    :raises RuntimeError: If unexpected backend errors occurred.
+
+    :examples:
+        >>> dem = np.array([[1, 1, 1],
+        ...                 [1, 2, 1],
+        ...                 [1, 1, 1]], dtype="float32")
+        >>> coeffs = get_quadric_coefficients(dem, resolution=1.0)
+        >>> coeffs.shape
+        (9, 3, 3)
+        >>> coeffs[:, 1, 1]
+        array([ 1.,  0.,  0., -1., -1.,  0.,  0.,  0.,  2.])
+
+    :returns: An array of coefficients for each pixel of shape (9, row, col).
+    """
+    # This function only formats and validates the inputs. For the true functionality, see _get_quadric_coefficients()
+    dem_arr = xdem.spatial_tools.get_array_and_mask(dem)[0]
+
+    if len(dem_arr.shape) != 2:
+        raise ValueError(
+            f"Invalid input array shape: {dem.shape}, parsed into {dem_arr.shape}. "
+            "Expected 2D array or 3D array of shape (1, row, col)"
+        )
+
+    if any(dim < 3 for dim in dem_arr.shape):
+        raise ValueError(f"DEM (shape: {dem.shape}) is too small. Smallest supported shape is (3, 3)")
+
+    # Resolution is in other tools accepted as a tuple. Here, it must be just one number, so it's best to sanity check.
+    if isinstance(resolution, Sized):
+        raise ValueError("Resolution must be the same for X and Y directions")
+
+    # Try to run the numba JIT code. It should never fail at this point, so if it does, it should be reported!
+    try:
+        coeffs = _get_quadric_coefficients(dem_arr, resolution)
+    except Exception as exception:
+        raise RuntimeError("Unhandled numba exception. Please raise an issue of what happened.") from exception
+
+    return _get_quadric_coefficients(dem_arr, resolution)
+
+
+@numba.njit(parallel=True)
+def _get_quadric_coefficients(
+    dem: np.ndarray,
+    resolution: float,
+) -> np.ndarray:
+    """
+    Run the pixel-wise analysis in parallel.
+
+    See the xdem.terrain.get_quadric_coefficients() docstring for more info.
+    """
+    # Rename the resolution to be consistent with the ArcGIS reference.
+    L = resolution
+
+    # Allocate the output.
+    output = np.empty((9,) + dem.shape, dtype=dem.dtype) + np.nan
+
+    # Loop over every pixel concurrently.
+    for i in numba.prange(dem.size):
+        # Derive its associated row and column index.
+        col = i % dem.shape[1]
+        row = int(i / dem.shape[1])
+
+        # Extract the pixel and its 8 immediate neighbours.
+        # If the border is reached, just duplicate the closest neighbour to obtain 9 values.
```

Comment:
Same as comment below for the edge behaviour

Reply:
How would one otherwise do it with edges? I'm very new to this!

Reply:
For consistency maybe the exact same as above, considering the edge as NaNs? Not sure how `fill_nearest` would work on an exact grid.
And maybe `default` should be more explicit `nofill` or `expand_nodata`... (not super inspired)?

Reply:
Added in e940f80. No tests yet though...